adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1395 @@
+	RewardCriterion   string      `json:"rewardCriterion"`
+	ScreenOrientation string      `json:"screenOrientation"`
+	PrivacyUrl        string      `json:"privacyUrl"`
+}
+
+type ImageInfo struct {
+	Url       string `json:"url"`
+	Height    int64  `json:"height"`
+	FileSize  int64  `json:"fileSize"`
+	Sha256    string `json:"sha256"`
+	ImageType string `json:"imageType"`
+	Width     int64  `json:"width"`
+}
+
+type Icon struct {
+	Url       string `json:"url"`
+	Height    int64  `json:"height"`
+	FileSize  int64  `json:"fileSize"`
+	Sha256    string `json:"sha256"`
+	ImageType string `json:"imageType"`
+	Width     int64  `json:"width"`
+}
+
+type VideoInfo struct {
+	VideoDownloadUrl string  `json:"videoDownloadUrl"`
+	VideoDuration    int32   `json:"videoDuration"`
+	VideoFileSize    int32   `json:"videoFileSize"`
+	Sha256           string  `json:"sha256"`
+	VideoRatio       float32 `json:"videoRatio"`
+	Width            int32   `json:"width"`
+	Height           int32   `json:"height"`
+}
+
+type ApkInfo struct {
+	Url           string       `json:"url"`
+	FileSize      int64        `json:"fileSize"`
+	Sha256        string       `json:"sha256"`
+	PackageName   string       `json:"packageName"`
+	SecondUrl     string       `json:"secondUrl"`
+	AppName       string       `json:"appName"`
+	Permissions   []Permission `json:"permissions"`
+	VersionName   string       `json:"versionName"`
+	AppDesc       string       `json:"appDesc"`
+	AppIcon       string       `json:"appIcon"`
+	DeveloperName string       `json:"developerName"`
+}
+
+type MediaFile struct {
+	Mime     string `json:"mime"`
+	Width    int64  `json:"width"`
+	Height   int64  `json:"height"`
+	FileSize int64  `json:"fileSize"`
+	Url      string `json:"url"`
+	Sha256   string `json:"sha256"`
+}
+
+type Monitor struct {
+	EventType string   `json:"eventType"`
+	Url       []string `json:"url"`
+}
+
+type Permission struct {
+	PermissionLabel string `json:"permissionLabel"`
+	GroupDesc       string `json:"groupDesc"`
+	TargetSDK       string `json:"targetSDK"`
+}
+
+type RewardItem struct {
+	Type   string `json:"type"`
+	Amount int32  `json:"amount"`
+}
+
+type adapter struct {
+	endpoint  string
+	extraInfo ExtraInfo
+}
+
+type ExtraInfo struct {
+	TrackingUrl string `json:"trackingUrl,omitempty"`
+}
+
+func (a *adapter) MakeRequests(openRTBRequest *openrtb2.BidRequest,
+	reqInfo *adapters.ExtraRequestInfo) (requestsToBidder []*adapters.RequestData, errs []error) {
+	// the upstream code already confirms that there is a non-zero number of impressions
+	numRequests := len(openRTBRequest.Imp)
+	var request HuaweiAdsRequest
+	var header http.Header
+	var multislot = make([]Adslot30, 0, numRequests)
+
+	var huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds
+	var err1 error
+	for index := 0; index < numRequests && numRequests > 0; index++ {
+		huaweiAdsImpExt, err1 = unmarshalExtImpHuaweiAds(&openRTBRequest.Imp[index])
+		if err1 != nil || huaweiAdsImpExt == nil {
+			errs = append(errs, err1)
```

Comment:
You probably should set `err1` to be non nil in the case that `huaweiAdsImpExt` is nil.

Reply:
corrected